File: yep/src/main/java/catchla/yep/view/FloatingActionMenu.java
package catchla.yep.view;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.content.Context;
import android.support.design.widget.AnimationUtilsTrojan;
import android.support.design.widget.FloatingActionButtonImplTrojan;
import android.support.v4.view.ViewCompat;
import android.support.v7.widget.CardView;
import android.util.AttributeSet;
import android.view.View;

/**
 * Created by mariotaku on 15/12/14.
 */
public class FloatingActionMenu extends CardView {
    private boolean mIsHiding;

    public FloatingActionMenu(final Context context) {
        super(context);
    }

    public FloatingActionMenu(final Context context, final AttributeSet attrs) {
        super(context, attrs);
    }

    public FloatingActionMenu(final Context context, final AttributeSet attrs, final int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public void hide() {
        hide(null);
    }

    public void show() {
        show(null);
    }

    public void show(final InternalVisibilityChangedListener listener) {
        if (this.getVisibility() != View.VISIBLE) {
            if (ViewCompat.isLaidOut(this) && !this.isInEditMode()) {
                this.setAlpha(0f);
                this.setScaleY(0f);
                this.setScaleX(0f);
                this.animate()
                        .scaleX(1f)
                        .scaleY(1f)
                        .alpha(1f)
                        .setDuration(FloatingActionButtonImplTrojan.SHOW_HIDE_ANIM_DURATION)
                        .setInterpolator(AnimationUtilsTrojan.FAST_OUT_SLOW_IN_INTERPOLATOR)
                        .setListener(new AnimatorListenerAdapter() {
                            @Override
                            public void onAnimationStart(Animator animation) {
                                FloatingActionMenu.this.setVisibility(View.VISIBLE);
                            }

                            @Override
                            public void onAnimationEnd(Animator animation) {
                                if (listener != null) {
                                    listener.onShown();
                                }
                            }
                        });
            } else {
                this.setVisibility(View.VISIBLE);
                this.setAlpha(1f);
                this.setScaleY(1f);
                this.setScaleX(1f);
                if (listener != null) {
                    listener.onShown();
                }
            }
        }
    }

    public void hide(final InternalVisibilityChangedListener listener) {
        if (mIsHiding || getVisibility() != View.VISIBLE) {
            // A hide animation is in progress, or we're already hidden. Skip the call
            if (listener != null) {
                listener.onHidden();
            }
            return;
        }

        if (!ViewCompat.isLaidOut(this) || isInEditMode()) {
            // If the view isn't laid out, or we're in the editor, don't run the animation
            this.setVisibility(View.GONE);
            if (listener != null) {
                listener.onHidden();
            }
        } else {
            animate()
                    .scaleX(0f)
                    .scaleY(0f)
                    .alpha(0f)
                    .setDuration(FloatingActionButtonImplTrojan.SHOW_HIDE_ANIM_DURATION)
                    .setInterpolator(AnimationUtilsTrojan.FAST_OUT_SLOW_IN_INTERPOLATOR)
                    .setListener(new AnimatorListenerAdapter() {
                        @Override
                        public void onAnimationStart(Animator animation) {
                            mIsHiding = true;
                            FloatingActionMenu.this.setVisibility(View.VISIBLE);
                        }

                        @Override
                        public void onAnimationCancel(Animator animation) {
                            mIsHiding = false;
                        }

                        @Override
                        public void onAnimationEnd(Animator animation) {
                            mIsHiding = false;
                            FloatingActionMenu.this.setVisibility(View.GONE);
                            if (listener != null) {
                                listener.onHidden();
                            }
                        }
                    });
        }
    }


    interface InternalVisibilityChangedListener {
        void onShown();

        void onHidden();
    }
}

File: src/main/java/catchla/yep/view/FloatingActionMenu.kt
package catchla.yep.view

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.content.Context
import android.support.design.widget.AnimationUtilsAccessor
import android.support.design.widget.FloatingActionButtonImplAccessor
import android.support.v4.view.ViewCompat
import android.support.v7.widget.CardView
import android.util.AttributeSet
import android.view.View

/**
 * Created by mariotaku on 15/12/14.
 */
class FloatingActionMenu : CardView {
    private var hiding: Boolean = false

    constructor(context: Context) : super(context) {
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
    }

    @JvmOverloads fun show(listener: InternalVisibilityChangedListener? = null) {
        if (this.visibility != View.VISIBLE) {
            if (ViewCompat.isLaidOut(this) && !this.isInEditMode) {
                this.alpha = 0f
                this.scaleY = 0f
                this.scaleX = 0f
                this.animate().scaleX(1f).scaleY(1f).alpha(1f).setDuration(FloatingActionButtonImplAccessor.SHOW_HIDE_ANIM_DURATION.toLong()).setInterpolator(AnimationUtilsAccessor.FAST_OUT_SLOW_IN_INTERPOLATOR).setListener(object : AnimatorListenerAdapter() {
                    override fun onAnimationStart(animation: Animator) {
                        this@FloatingActionMenu.visibility = View.VISIBLE
                    }

                    override fun onAnimationEnd(animation: Animator) {
                        listener?.onShown()
                    }
                })
            } else {
                this.visibility = View.VISIBLE
                this.alpha = 1f
                this.scaleY = 1f
                this.scaleX = 1f
                listener?.onShown()
            }
        }
    }

    @JvmOverloads fun hide(listener: InternalVisibilityChangedListener? = null) {
        if (hiding || visibility != View.VISIBLE) {
            // A hide animation is in progress, or we're already hidden. Skip the call
            listener?.onHidden()
            return
        }

        if (!ViewCompat.isLaidOut(this) || isInEditMode) {
            // If the view isn't laid out, or we're in the editor, don't run the animation
            this.visibility = View.GONE
            listener?.onHidden()
        } else {
            animate().scaleX(0f).scaleY(0f).alpha(0f).setDuration(FloatingActionButtonImplAccessor.SHOW_HIDE_ANIM_DURATION.toLong()).setInterpolator(AnimationUtilsAccessor.FAST_OUT_SLOW_IN_INTERPOLATOR).setListener(object : AnimatorListenerAdapter() {
                override fun onAnimationStart(animation: Animator) {
                    hiding = true
                    this@FloatingActionMenu.visibility = View.VISIBLE
                }

                override fun onAnimationCancel(animation: Animator) {
                    hiding = false
                }

                override fun onAnimationEnd(animation: Animator) {
                    hiding = false
                    this@FloatingActionMenu.visibility = View.GONE
                    listener?.onHidden()
                }
            })
        }
    }


    interface InternalVisibilityChangedListener {
        fun onShown()

        fun onHidden()
    }
}
